text/0000-toolchains.md

```diff
@@ -0,0 +1,163 @@
+
+## Pinning
+
+The `"platform"` section of `package.json` selects the platform image associated with a package. This is a change from earlier versions of Notion, which used the key `"toolchain"`, since the toolchain consists of _both_ the platform image _and_ the pinned package binaries (see above).
+
+Users can pin the platform by manually editing the `package.json` `"platform"` section or via [`notion pin`](https://github.com/notion-cli/rfcs/pull/24).
+
+## Installation
+
+The `notion install` command installs a tool to the user toolchain.
+
+When installing a version of the Node runtime, Notion also installs the default version of npm bundled with that verison of Node. This can be overridden with a subsequent `notion install npm` command.
+
+When installing a package binary to the user toolchain, Notion checks the package for a `"platform"` key to pin the user tool to a specific platform image. If there is no `"platform"` key in the package manifest, it defaults to the current platform image configured in the user toolchain. If there is no current platform image, `notion install` fails with an error message suggesting the user choose at least a Node runtime version.
+
+# Critique
+[critique]: #critique
+
+It's natural to question whether pinning for reproducibility is worth the cost of extra fetching. An alternative approach would be to allow projects to specify less precise version requirements (such as the ranges typically expressed in the `"engines"` field of `package.json`) and assume most differences will be benign. However, behavioral divergences between versions of Node do happen and are tricky bugs to nail down. Putting in extra work up front to ensure that these divergences cannot happen, by construction should pay dividends when scaled across the Node ecosystem. And over time, we can investigate optimization techniques to save time and disk space for fetching multiple similar versions.
+
+Theoretically, it might make more sense to put the `"platform"` section in a lockfile. But since there isn't a standardized single lockfile format for JS, and those formats aren't extensible, and we don't want to impose a whole new file to add to JS projects, using the package manifest seemed like the least imposition on users.
+
+Another reasonable criticism is that pinning the Node version for tools in the user toolchain means that users will not automatically benefit from performance and security improvements in Node. There are a couple of reasons this is outweighed by the benefits of pinning. First, as users upgrade the tool version itself, they will automatically get re-pinned to the newer version of Node specified by newer versions of the tool's `package.json`. Second, we should at least allow users the option to override the tool's specified Node version. But by letting the tool choose its platform version by default, users have a stronger guarantee that the tool will work and continue to work consistently based on how it was tested.
```

Comment:
Can we add the upgrading-a-tool-updates-its-Node-version assumption to the Installation section above so it isn't just embedded here?  It would also be nice to specify that as user can update the Node version of a binary by "reinstalling" its package, even if the version doesn't change.

Reply:
Updated!